<a href="https://colab.research.google.com/github/Khushitz/MACHINE-LEARNING/blob/MACHINE-LEARNING/test_train_validationsplit.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
from sklearn.model_selection import train_test_split

In [4]:
x = np.array([[1], [2], [3], [4], [5], [6], [7], [8], [9], [10]])
y = np.array([2, 4, 6, 8, 10, 12, 14, 16, 18, 20])


In [9]:
# Step 1: Split into Train (70%) and Temp (30%)

x_train,x_temp,y_train,y_temp = train_test_split(x,y,test_size=0.3, random_state=42)

In [10]:
# Step 2: Split Temp into Validation (15%) and Test (15%)
x_val,x_test,y_val,y_test = train_test_split(x_temp,y_temp,test_size=0.5,random_state=42)

In [11]:
print("training size:", len(x_train))
print("validation size:", len(x_val))
print("test_size:",len(x_test))

training size: 7
validation size: 1
test_size: 2


| Dataset    | Percentage | Purpose       |
| ---------- | ---------- | ------------- |
| Train      | 70%        | Learn pattern |
| Validation | 15%        | Tune model    |
| Test       | 15%        | Final check   |


In [12]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import OneHotEncoder
from sklearn.metrics import r2_score


In [13]:
data = pd.DataFrame({
    'Area': [800, 1200, 1500, 2000, 2200, 2500, 3000, 3500],
    'Location': ['Rural', 'Urban', 'Urban', 'SemiUrban', 'Urban', 'Rural', 'SemiUrban', 'Urban'],
    'Price': [2000000, 5000000, 6500000, 7000000, 9000000, 3000000, 7500000, 12000000]
})
print("Dataset:\n", data)

Dataset:
    Area   Location     Price
0   800      Rural   2000000
1  1200      Urban   5000000
2  1500      Urban   6500000
3  2000  SemiUrban   7000000
4  2200      Urban   9000000
5  2500      Rural   3000000
6  3000  SemiUrban   7500000
7  3500      Urban  12000000


In [15]:
#one hot encoding
encoder = OneHotEncoder(drop='first',sparse_output=False)
location_encoded = encoder.fit_transform(data[['Location']])
encoded_df=pd.DataFrame(location_encoded,columns=encoder.get_feature_names_out())

In [16]:
#combine features
x=pd.concat([data[['Area']],encoded_df], axis=1)
y=data['Price']


In [17]:
#train-test split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.25,random_state=42)

The test_size=0.25 argument specifies that 25% of the data will be allocated for the testing set, leaving 75% for the training set.
The random_state=42 argument ensures that the data is shuffled and split in the exact same way every time the code runs, which is crucial for reproducibility of results.


In [18]:
#train model
model =LinearRegression()
model.fit(x_train,y_train)

LinearRegression()

In [20]:
#evaluate
y_pred = model.predict(x_test)
#This step is crucial for evaluating how well the trained model performs on data it has not seen during training.

print("\nR2 Score:", r2_score(y_test,y_pred))
print("Intercept:",model.intercept_)
print("Coefficients:",model.coef_)


R2 Score: -4.113935682508677
Intercept: 187499.99999998976
Coefficients: [2.26562500e+03 1.39843750e+06 3.54166667e+06]


R2 Score: -4.113935682508677 (indicates a poor fit, as a good R2 score is typically between 0 and 1)
Intercept: 187499.99999998976
Coefficients: [2.26562500e+03, 1.39843750e+06, 3.54166667e+06] (suggests a model with three features)
